core/domain/user_services.py

```diff
@@ -2287,3 +2291,109 @@ def get_dashboard_stats(user_id):
         'num_ratings': num_ratings,
         'average_ratings': average_ratings
     }
+
+
+def _get_checkpoints_in_order(init_state_name, states):
+    """Returns the checkpoints of an exoloration in sequential order.
+
+    Args:
+        init_state_name: str. The name of the first state of the exploration.
+        states: dict(state). All states of the exploration.
+
+    Returns:
+        list(str). List of all checkpoints of the exploration in sequential
+        order.
+    """
+    queue = []
+    queue.append(init_state_name)
+    checkpoint_state_names = []
+    visited_state_names = []
+    while len(queue) > 0:
+        current_state_name = queue.pop()
+        if current_state_name not in visited_state_names:
+            visited_state_names.append(current_state_name)
+            current_state = states[current_state_name]
+            if (current_state.card_is_checkpoint and (
+                current_state_name not in checkpoint_state_names)):
+                checkpoint_state_names.append(current_state_name)
+            for answer_group in current_state.interaction:
+                queue.append(answer_group.outcome.dest)
+
+    return checkpoint_state_names
+
+
+def update_learner_checkpoint_progress(
+    user_id, exploration_id, state_name, exp_version
+):
+    """Sets the furthest reached and most recently reached checkpoint in
+    an exploration by the user.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+        state_name: str. The state name of the most recently reached checkpoint.
+        exp_version: int. The exploration version of the most recently reached
+            checkpoint.
+    """
+
+    exploration_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+    if exploration_user_model is None:
+        exploration_user_model = user_models.ExplorationUserDataModel.create(
+            user_id, exploration_id)
+
+    latest_exploration = exp_fetchers.get_exploration_by_id(exploration_id)
```

Comment:
Why are we getting the latest version? Should it not be the version that the function receives as parameter ?

Reply:
Updated it to fetch the `current_exploration` user has interacted with using the exploration id provided by the url endpoint.